pkg/retry/retry.go

```diff
@@ -141,52 +146,61 @@ func (s *Server) handleRetry() http.Handler {
-
-				if err := s.github.RedeliverEvent(ctx, *event.ID); err != nil {
-					var acceptedErr *github.AcceptedError
-					if errors.As(err, &acceptedErr) {
-						logger.Infow("skipping redeliver event because it has already been submitted to GitHub",
-							"eventID", *event.ID,
-							"guid", *event.GUID,
-							"error", err,
-						)
-						continue
-					}
-
-					// found an unaccepted error, check if its already in the events table
-					exists, err := s.datastore.DeliveryEventExists(ctx, s.eventsTableID, *event.GUID)
-					if err != nil {
-						logger.Errorw("failed to call BigQuery",
-							"method", "DeliveryEventExists",
-							"code", http.StatusInternalServerError,
-							"body", errDeliveryEventExists,
-							"error", err,
-						)
-						http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
-						return
-					}
-
-					if exists {
-						logger.Infow("skipping redeliver event because it has already been processed", "eventID", *event.ID, "guid", *event.GUID)
-						continue
-					}
-
-					logger.Errorw("failed to call RedeliverEvent, stop processing",
+				failedEventsHistory = append(failedEventsHistory, &EventIdentifier{eventID: event.ID, guid: *event.GUID})
+			}
+		}
+
+		// work backwards from the list of failed events then attempt redelivery and
+		// increment the newCheckpoint in an effort to close the gap to the most
+		// recent event, this should alleviate pressure on future runs
+		for i := len(failedEventsHistory) - 1; len(failedEventsHistory) > 0 && i >= 0; i-- {
+			eventIdentifier := failedEventsHistory[i]
+			logger.Infow("redeliver failed event", "eventID", eventIdentifier.eventID)
+
+			if err := s.github.RedeliverEvent(ctx, *eventIdentifier.eventID); err != nil {
+				var acceptedErr *github.AcceptedError
+				if errors.As(err, &acceptedErr) {
+					logger.Infow("skipping redeliver event because it has already been submitted to GitHub",
+						"eventID", *eventIdentifier.eventID,
+						"guid", eventIdentifier.guid,
+						"error", err,
+					)
+					continue
+				}
+
+				// found an unaccepted error, check if its already in the events table
+				exists, err := s.datastore.DeliveryEventExists(ctx, s.eventsTableID, eventIdentifier.guid)
+				if err != nil {
+					logger.Errorw("failed to call BigQuery",
+						"method", "DeliveryEventExists",
 						"code", http.StatusInternalServerError,
-						"body", errCallingGitHub,
-						"method", "RedeliverEvent",
-						"event", *event,
-						"guid", *event.GUID,
+						"body", errDeliveryEventExists,
 						"error", err,
-						"totalEventCount", totalEventCount,
-						"failedEventCount", failedEventCount,
 					)
 					http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
 					return
 				}
+
+				if exists {
+					logger.Infow("skipping redeliver event because it has already been processed", "eventID", *eventIdentifier.eventID, "guid", eventIdentifier.guid)
```

Comment:
```suggestion
					logger.Infow("skipping redeliver event because it has already been processed", "eventID", eventIdentifier.eventID, "guid", eventIdentifier.guid)
```

Reply:
OK, I didn't pay attention to the value we got from GH. Consider de-pointer them in the EventIdentifier to make the code here (and other places) cleaner